src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala

```diff
@@ -5,67 +5,229 @@
+        }
+      }.getOrElse {
+        trace("Cannot find old name of symbol reliably; attempting fallback")
+        try {
+          selected match {
+            case Import(_, List(selector)) => selector.rename.toString
+            case _ => selected match {
+              case s: Select => s.name.decoded
+              case other => other.nameString
+            }
+          }
+        } catch {
+          case NonFatal(e) =>
+            trace("Cannot find old name of symbol; giving up")
+            "<no-name>"
+        }
+      }
+    } \\ { name =>
+      trace(s"SingleTreeSelection.name: $name")
+    }
+  }
 
-    def occurrencesForSymbol(s: Symbol) = {
-      val occurrences = index.occurences(s)
-      occurrences map (_.namePosition) filter  (_.isRange)
+  protected def findOccurrences(selection: SingleTreeSelection): List[(RangePosition, Tree)] = {
+    /*
+     * A lazy val is represented by a ValDef and an associated DefDef that contains the initialization code.
+     * Unfortunately, the Scala compiler does not set modifier positions for the DefDef, which is a problem for us,
+     * because we are relying on the DefDef when printing out source code. The following function patches the affected
+     * DefDefs accordingly.
+     *
+     * See #1002392 and #1002502
+     */
+    def eventuallyFixModifierPositionsForLazyVals(t: Tree): Unit = t match {
+      case dd: DefDef if dd.mods.isLazy && dd.mods.positions.isEmpty =>
+        val vd = selection.root.find {
+          case vd: ValDef if vd.mods.isLazy && !vd.mods.positions.isEmpty && dd.pos.point == vd.pos.point => true
+          case _ => false
+        }
+
+        vd.foreach { vd =>
+          // Note that we patch the DefDef in place because we need the correctly set positions in the tree
+          // later in the refactoring process. A precursor of this code pretended to be pure, by copying
+          // dd and then calling 'mods.setPositions' on the copy. This had exactly the same (needed) side effects
+          // however, as the Modifier object affected by setPositions was the same anyway. If you are interested
+          // in the related discussion, take a look at https://github.com/scala-ide/scala-refactoring/pull/78.
+          dd.mods.setPositions(vd.asInstanceOf[ValDef].mods.positions)
+        }
+
+      case _ => ()
     }
 
-    val selectedTree = (new FileSelection(file, global.unitOfFile(file).body, from, to)).findSelectedWithPredicate {
-      case (_: global.TypeTree) | (_: global.SymTree) | (_: global.Ident) => true
-      case _ => false
+    /*
+     * ImportSelectorTrees need special treatment, since it is not a priory clear which side (the name or rename side) of
+     * the tree is relevant for us.
+     */
+    def positionsWithPotentiallyReferencingTrees(trees: Seq[Tree]): Seq[(Position, Tree)] = {
+      trees.flatMap {
+        case ImportSelectorTree(name, rename) => Seq(name.namePosition() -> name, rename.namePosition() -> rename)
```

Comment:
just a question, is it better to work with `ImportSelectorTrees` or pure `Imports`? Honestly I don't work with enriched trees and looks that I'm a bit reckless.